### DISCUSSION TITLE
Integrity check when loading WASM

### DISCUSSION THREAD

**Discussion by matthiasgeihs** [wasm-bindgen/wasm-bindgen#3923] [Q&A] [answered]
We are loading a WASM file via the generated `init` and `initSync` functions.

**Question:**
Is there any integrated integrity check on the provided WASM file? Does the loader check that the expected WASM is provided? (e.g., verify against target hash)

**Comment by daxpedda** [comment]
`wasm-bindgen` does nothing in this regard, no.

**Reply by matthiasgeihs** [reply]
I’m wondering: has this been considered? What are the reasons for doing it / not doing it?

**Reply by daxpedda** [reply]
It has not.
I wouldn't know what exactly the benefits would be.

For safety, TCP, encryption, compression algorithm and the like already do their own checksum calculations.

For security, if an attacker can modify the delivered Wasm module, they can modify the checksum as well, so I don't see a big benefit here. Apart from the fact that if somebody breaks your HTTPS all hell breaks loose already, the Wasm module would be your least worry.

**Reply by matthiasgeihs** [reply]
You are assuming that javascript glue code and WASM come from the same source. They may not. Consider the case where the client glue code is distributed ahead of time via a trusted package manager. Due to size constraints, the WASM, however, is only loaded at runtime from a potentially untrusted source. In this case, it would be healthy to check the integrity of the WASM before loading it. Wouldn't it?

**Reply by daxpedda** [reply, marked as answer]
Yes, the assumption here was of course is that you are able to trust your source.
If you can't trust your source, something like that might make sense indeed.

Though I don't expect this to make any sense in `wasm-bindgen`, this can easily be done externally.

**Reply by matthiasgeihs** [reply]
Yep, that's what I thought. Thanks for explaining also the context!

**Reply by yawn** [reply]
Discussion is closed / answered but I'd like to quickly chime in here with a related item: Subresource Integrity (SRI) currently supports only a limited set of resources (no WASM). Would you be open to a discussion on changing `__wbg_load` to accommodate some extension point or optional content verification method that users could implement? Currently doing this requires overriding the entire method since (to the best of my knowledge) hashing and / or signature verification is in conflict with `instantiateStreaming`.

The basic argument in favour of this would be the same logic that applies to SRI (quote from the spec):

> Sites and applications on the web are rarely composed of resources from only a single origin. For example, authors pull scripts and styles from a wide variety of services and content delivery networks, and must trust that the delivered representation is, in fact, what they expected to load. If an attacker can trick a user into downloading content from a hostile server (via DNS [RFC1035] poisoning, or other such means), the author has no recourse. Likewise, an attacker who can replace the file on the Content Delivery Network (CDN) server has the ability to inject arbitrary content.

If we had a way to verify WASM content before loading we add SRI to the loader and be done with it.
